### ISSUE TITLE
Slack replying to IRC stopped working

### ISSUE THREAD

**Issue by k3ilbasa** [ekmartin/slack-irc#30]
Any ideas why I would get IRC messages into slack, but my slack replies aren't being sent?  It was working initially when I set it up, are there any logs I can upload or anything?

**Comment by k3ilbasa** [comment]
Could this possibly have something to do with the bot on slack showing up as offline?

**Comment by andebor** [comment]
Maybe try to show the debug log by setting `NODE_ENV=development`. There might be some useful info there. (btw, more on this in the readme)

**Comment by ekmartin** [comment]
Does this happen even after a restart of the bot? Are you sure the bot's actually present in the Slack channel? If it is, try @andebor suggestion :)

**Comment by k3ilbasa** [comment]
I havent tried setting NODE_ENV=development yet, but I will do that next.  I did restart the bot and it started working again.  Interestingly enough, when I restarted the bot on my ubuntu server, the IRCBot I created in Slack showed as online again.  Why would this happen, what could cause it to show offline?

**Comment by ekmartin** [comment]
There might be a problem with node-slack and reconnecting after it loses connection, since a restart of the bot fixed it. I'll investigate later. When it comes to the offline/online thing I'm not sure, but it might be that Slack shows it as offline as long as there's no active WebSocket connections to the bot.

**Comment by ekmartin** [comment]
Might be related to https://github.com/slackhq/node-slack-client/issues/5#issuecomment-71195036

**Comment by voneiden** [comment]
I encountered a similar issue today, I've enabled the debug log in case the problem comes back again so I can drop some more details. It worked fine for 7 days so it's certainly not a setup issue. Messages were being delivered from IRC to Slack but not the other way. Also running on a raspi.

**Comment by voneiden** [comment]
So the bot does indeed appear now offline on Slack, but does still deliver messages from IRC to Slack. Messages from Slack are not picked up however. There's nothing useful in the debug log. No mention of disconnects or anything along the lines of hinting something went wrong.

**Comment by mrsmn** [comment]
Hi there,

I'm affected by the same issue: after a random amount of time, the messages from Slack to IRC stop working while the opposite way still works neatly. Nothing relevant in my logs either.

Cheers.

**Comment by ekmartin** [comment]
Hm, really weird. You don't see anything about disconnecting/reconnecting in the logs either? This is how it looks (with `NODE_ENV=development`) for me if the bot loses connection:
```
error: Received error event from Slack ENOTFOUND
[Mon May 25 2015 16:19:10 GMT+0200 (CEST)] INFO Reconnecting in 14000ms
[Mon May 25 2015 16:19:24 GMT+0200 (CEST)] INFO Attempting reconnect
[Mon May 25 2015 16:19:24 GMT+0200 (CEST)] INFO Connecting...
debug: Connected to Slack
```

Since the IRC bot is still connected all messages from Slack to IRC should at least have been logged, so the issue most likely is with the WebSocket client not managing to reconnect correctly. Messages sent from IRC to Slack are sent with the web API, and not the real time messaging API, so this is probably why IRC -> Slack still works (but not Slack -> IRC).

Sounds a lot similar to https://github.com/slackhq/node-slack-client/issues/5#issuecomment-99922443.

**Comment by ekmartin** [comment]
You can also expose some more debug information by doing `export SLACK_LOG_LEVEL=debug` before running slack-irc. This'll show pings/pongs too.

**Comment by mrsmn** [comment]
Hi there,

I already had set ```SLACK_LOG_LEVEL=debug``` so all I see in the logs is ```25 May 14:29:45 - SEND: PONG irc.example.com```. Sadly, nothing like you encounter when the bot loses connection.

Cheers.

**Comment by ekmartin** [comment]
That looks like a pong event from the IRC server, for node-slack-client they seem to look like this:
```
[Mon May 25 2015 16:35:13 GMT+0200 (CEST)] DEBUG ping
[Mon May 25 2015 16:35:14 GMT+0200 (CEST)] DEBUG pong
```

Do you have any of those in the log (after the messages stopped)?

**Comment by mrsmn** [comment]
Didn't have any node-slack-client debug logs so I indeed had to export ```SLACK_LOG_LEVEL=debug``` and now it works as expected:

```
[Mon May 25 2015 14:48:04 GMT+0000 (UTC)] DEBUG ping
[Mon May 25 2015 14:48:04 GMT+0000 (UTC)] DEBUG pong
```

Now I should have all the logs available if the bot stops again.

**Comment by voneiden** [comment]
I've also added ``export SLACK_LOG_LEVEL=debug``. Will report back when the issue appears again. Last time it took 15 days, so it might take a while.

**Comment by ekmartin** [comment]
Thanks, I'm doing the same on one of my servers. Not sure if you're referring to me or @mrsmn, @voneiden, but I don't think I've seen it happen on my cheap DigitalOcean droplet at least (512 MB RAM).

**Comment by voneiden** [comment]
Oh right, two Martins, sorry. Meant @mrsmn. The "maybe related"  slackhq/node-slack-client#5 you mentioned earlier had a bunch of people with raspberry pi's having issues so I'm curious if slow devices could be the cause for this one too. 

DigitalOcean seems to be using SSD's and 2-3 GHz processors, while the raspi is stuck with SD cards and somewhere around 700 MHz.

**Comment by mrsmn** [comment]
@voneiden Personally, slack-irc is running on a Xeon E3-1245V2 with 32GB of RAM so I don't think host performance could be the issue.

**Comment by andebor** [comment]
I think this might have something to do with our problem
```
[Tue May 26 2015 18:24:22 GMT+0000 (UTC)] DEBUG Unknown message type: team_migration_started
[Tue May 26 2015 18:24:22 GMT+0000 (UTC)] DEBUG { type: 'team_migration_started' }
```
After this event, slack stopped sending messages to irc, while messages from irc still got posted to slack.

After seeing [this](https://api.slack.com/events/team_migration_started), I think it might be a reconnect issue. This also explains why this is not a persistent issue for everyone.

**Comment by ekmartin** [comment]
That sounds quite reasonable @andebor, good find! Can't seem to find any previous node-slack-client issues for it, maybe you should make one?

**Comment by andebor** [comment]
I'll make an issue, @ekmartin. I'd just like to see if I can replicate the results again to confirm the issue. Hopefully it won't be too long before it happens again. If anyone else notices the same event, please let me know.

**Comment by andebor** [comment]
After about 500k lines of useless logs, it finally (sort of) happened again today. At least it got disconnected. There were literally no entries regarding the disconnect in the logs, but I suspect that it's caused by the same event. As far as I can tell, there is nothing handling those events in the node-slack-client, so I reported the issue (slackhq/node-slack-client#51).

It would've been nice if there were any way to check when your team has been through a team migration. Then we would've been able to compare the timestamps at least.

*If anyone have any log entries even remotely related to their disconnects, please post them here.*

**Comment by ulion** [comment]
I encountered this too. I try to workaround it by call reconnect or login in slack on close handler, but which caused other problem: messages from Slack to IRC get double and triple, etc. any idea why?

**Comment by ulion** [comment]
currently, my workaround to this issue, is to call login in the slack on close event handler, certainly, need set autoReconnect of slack client to false, and also need fix the slack-client code where pong timeout, to just close the connection, not call reconnect if autoReconnect is false.

**Comment by jonjonsson** [comment]
Any hope this will be fixed soon?

@ulion can you expand on your workaround? Can you put your fix on pastebin or similar? 

Thanks!

**Comment by ekmartin** [comment]
Issues like https://github.com/slackhq/node-slack-client/issues/5 have been there for quite a long time now, as it doesn't seem like node-slack-client is very high up on the priority list for Slack's development team. A fork that gathers together some of the suggested pull requests, fixes issues and writes some tests could probably be good.

**Comment by t-my** [comment]
I'm running:
 
* FreeBSD 10.0-RELEASE-p7
* Node 4.2.1
* NPM 3.3.9

export SLACK_LOG_LEVEL=debug && export NODE_ENV=development && node index.js --config slack-irc.json



Nothing when sending message to slack. Works normally when sending to IRC.

**Comment by ekmartin** [comment]
@t-my - does it work initially, and then stops working? If it doesn't work at all, please create a separate issue and I'll try debugging it with you.

**Comment by ekmartin** [comment]
Just a quick suggestion, have you remembered to invite the Slack bot user into the channels you want it to listen on? Using the `/invite nick` commad in Slack. Should probably be a check in slack-irc to make sure that the bot is in all the channels provided in the mapping.

**Comment by ngtjah** [comment]
Same issue. After a little while I have one-way traffic, IRC cannot hear slack.. It seems to happen anywhere from once a day to once a week...  I'll turn on the ping logging and report back.

**Comment by ngtjah** [comment]
```
[Thu Nov 05 2015 20:41:02 GMT-0700 (MST)] DEBUG pong
[Thu Nov 05 2015 20:41:07 GMT-0700 (MST)] DEBUG ping
[Thu Nov 05 2015 20:41:08 GMT-0700 (MST)] DEBUG pong
[Thu Nov 05 2015 20:41:12 GMT-0700 (MST)] DEBUG ping
[Thu Nov 05 2015 20:41:12 GMT-0700 (MST)] DEBUG pong
[Thu Nov 05 2015 20:41:14 GMT-0700 (MST)] DEBUG Unknown message type: team_migration_started
[Thu Nov 05 2015 20:41:14 GMT-0700 (MST)] DEBUG { type: 'team_migration_started' }
```
After that no more ping pong

**Comment by ekmartin** [comment]
Saw you found the open node-slack-client [issue](https://github.com/slackhq/node-slack-client/issues/51) for it @ngtjah. There's an open pull request too, but doesn't seem like Slack is doing much maintaining of the project at the moment. Should probably create a fork that includes some of the open pull requests and rely on that instead.

**Comment by ekmartin** [comment]
https://github.com/slackhq/node-slack-client/pull/54 was just merged and published in version 1.5.0 of node-slack-client. slack-irc 3.6.2 uses this version, so hopefully that'll prevent this issue from occurring. I'm closing this for now, but please let me know if the disconnects continue to happen.
